Kernel/Process.cpp

```diff
@@ -295,11 +295,28 @@ Process::~Process()
     VERIFY(!m_alarm_timer);
 
     PerformanceManager::add_process_exit_event(*this);
+}
 
-    if (m_list_node.is_in_list())
-        processes().with_exclusive([&](auto& list) {
-            list.remove(*this);
-        });
+bool Process::unref() const
+{
+    bool did_destruct = false;
+
+    // NOTE: We need to obtain the process list lock before doing anything,
+    //       because otherwise someone might get in between us lowering the
+    //       refcount and acquiring the lock.
+    processes().with_exclusive([&](auto& list) {
+        auto new_ref_count = deref_base();
+        if (new_ref_count == 0) {
+            // We have no more users, time to remove ourselves from the list.
+            if (m_list_node.is_in_list())
+                list.remove(*const_cast<Process*>(this));
+
+            delete this;
+            did_destruct = true;
+        }
+    });
+
+    return did_destruct;
```

Comment:
This can be simplified:

```cpp
return processes().with_exclusive([&](auto&) {
    if (deref_base())
        return false;
    m_list_node.remove();
	delete this;
    return true;
});
```

Reply:
I'm not quite sure about moving `deref_base` directly into the `if` because it's kinda ambiguous what its return value means when you're reading the code, but I'll apply the other parts.